proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java

```diff
@@ -2,53 +2,108 @@
+
+import java.io.ByteArrayInputStream;
 import java.io.DataOutputStream;
 import java.io.IOException;
 import java.io.OutputStream;
 import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
 import java.util.ArrayList;
 import java.util.List;
 
 /**
- * [Square] Tape converter for Lists of strings.
+ * [Square] Tape converter for Lists of strings with LZ4 compression support.
  *
- * Binary format is length[2b], {utf8Len[2b], utfSeq}*
+ * Binary format is signature[4b] = 0x54415045, version[4b] = 0x00000001, length[4b], {utf8Len[4b], utfSeq}*
  *
  * @author Tim Schmidt (tim@wavefront.com).
+ * @author Vasily Vorontsov (vasily@wavefront.com)
+ *
  */
 public class TapeStringListConverter implements FileObjectQueue.Converter<List<String>> {
-  private static final TapeStringListConverter INSTANCE = new TapeStringListConverter();
+  private static final Logger logger = LogManager.getLogger(TapeStringListConverter.class.getCanonicalName());
+
+  private static final TapeStringListConverter INSTANCE_DEFAULT = new TapeStringListConverter(false);
+  private static final TapeStringListConverter INSTANCE_COMPRESSION_ENABLED = new TapeStringListConverter(true);
+  private static final Charset UTF8 = Charset.forName("UTF-8");
+  private static final int CONTAINER_SIGNATURE = 0x54415045;
+  private static final int CONTAINER_VERSION = 0x00000001;
+
+  private final boolean isCompressionEnabled;
 
-  private TapeStringListConverter() {
-    // Singleton
+  private TapeStringListConverter(boolean isCompressionEnabled) {
+    this.isCompressionEnabled = isCompressionEnabled;
   }
 
-  public static TapeStringListConverter get() {
-    return INSTANCE;
+  public static TapeStringListConverter getDefaultInstance() {
+    return INSTANCE_DEFAULT;
+  }
+
+  public static TapeStringListConverter getCompressionEnabledInstance() {
+    return INSTANCE_COMPRESSION_ENABLED;
   }
 
   @Override
   public List<String> from(byte[] bytes) throws IOException {
-    ByteBuffer in = ByteBuffer.wrap(bytes);
-    int count = in.getShort();
-    List<String> result = new ArrayList<>(count);
-    for (int i = 0; i < count; ++i) {
-      int len = in.getShort();
-      result.add(new String(bytes, in.position(), len));
-      in.position(in.position() + len);
+    try {
+      ByteBuffer in;
+      byte[] uncompressedData;
+      if (bytes.length > 2 && bytes[0] == (byte) 0x4C && bytes[1] == (byte) 0x5A) { // LZ block signature
+        uncompressedData = IOUtils.toByteArray(new LZ4BlockInputStream(new ByteArrayInputStream(bytes)));
+        in = ByteBuffer.wrap(uncompressedData);
+      } else {
+        uncompressedData = bytes;
+        in = ByteBuffer.wrap(bytes);
+      }
+      int signature = in.getInt();
+      int version = in.getInt();
+      if (signature != CONTAINER_SIGNATURE || version != CONTAINER_VERSION) {
+        logger.error("Unknown data format retrieved from tape (signature = " + signature + ", version = " + version);
```

Comment:
Our error messages have a key and should probably stick to one (at this point we should also have a list of error codes in the README.md)